c10/core/impl/LocalDispatchKeySet.h

```diff
@@ -113,10 +102,51 @@ class C10_API ExcludeDispatchKeyGuard {
  private:
   // A little micro-optimization to save us from tls_get_addr call
   // on destruction
-  PODLocalDispatchKeySet* tls_;
+  LocalDispatchKeySetWrapper tls_wrapper_;
   DispatchKeySet exclude_;
 };
 
+template <uint64_t exclude>
+class C10_API ExcludeDispatchKeyGuard_NoOverlap {
+ public:
+  // If our exclude set does not overlap with c10::default_excluded_set, we can
```

Comment:
Hmmm please correct me if I'm wrong: if I want to add a key into c10::default_excluded_set in the future, do I have to check all call sites of `ExcludeDispatchKeyGuard_NoOverlap` and make sure they still don't overlap? ;) If so this seems a bit hard to maintain compared to the previous version.

Reply:
I'm merging both paths into a single guard. (Where you declare if there is overlap) I could make it automatic, but there aren't that many guards and I don't want a change in default_excluded_set to silently regress perf by kicking us off the hot path.

Reply:
I guess, what is unclear to me even after reading the explanation here, is if I change `default_excluded_set` and this static assert fires, what the heck am I supposed to do in that case? Am I supposed to edit the `has_overlap` arguments (won't that regress perf?). But then I am confused what else I am supposed to do. Is this a case of there being duplicated information that has to be kept up to date in multiple places? If that's the case, can we clearly indicate where else things have to be updated (or make it so that an update is not necessary in the first place?)

Reply:
Yes, the idea is that you have to update `has_overlap`; effectively forcing a discussion in code review about whether it's ok to knock that guard off the hot path. I can deduce `has_overlap`, it just means we can have silent perf regressions.